### ISSUE TITLE
[Feature Request] Hextuples Parser

### ISSUE THREAD

**Issue by ashleysommer** [oxigraph/oxigraph#1033]
Hextuples is an RDF format used where highly performant and parallel parsing is required at the expense of verbosity and filesize.
It is considered to be even faster to parse than NQuads due to its use of NDJSON inside the file.
https://github.com/ontola/hextuples

Hextuples support was added to Python RDFLib in 2022 https://github.com/RDFLib/rdflib/blob/main/rdflib/plugins/parsers/hext.py

I believe it would be a very easy win to get the hextuples format in oxigraph. Much like ntripes and nquads, it can easily be parsed in a streaming manner, and would support parallel parsing.

**Comment by ashleysommer** [comment]
I've only mentioned parsing here, (because this is the feature I need for my usecase) but hextuples serialization would be useful too.

**Comment by Tpt** [comment]
Thank you for your interest in Oxigraph and this enhancement idea.

>  It is considered to be even faster to parse than NQuads due to its use of NDJSON inside the file.

In JS/Python this sounds likely because the JSON parser is fast and written in C or C++ whereas the NQuads parser would be written in JS and likely slower (no SIMD...). In Oxigraph this does not sound obvious because both the JSON and the NQuads parser would be written in Rust and can be optimized in similar ways. For example the new oxttl parser relies on [`memchr`](https://docs.rs/memchr) to get some SIMD acceleration. A correct parser will have to do most the same expensive operations (validate IRIs, properly unescape unicode escape sequences, allocate buffers for the output values...). 

Hence, if the goal is "fast RDF parsing in Rust" not sure if it's worth it. There are a lot of NQuads out there and time might be better spent optimizing the NQuads parser. If the goal is "there are hextuples files out there we need to load and are not available in other formats", supporting hextuples would make more sense imho. In which case are we?

Anyway, if the goal is "fast RDF parsing in Rust" I would love to see a prototype and a benchmark before committing to maintain it long term. And if it's indeed significantly faster than NQuads, I would be happy to include it.

**Comment by ashleysommer** [comment]
You're absolutely right that in the case of a highly optimised compiled language like Rust, the benefits of JSON in Hextuples are diminished.

> if the goal is "fast RDF parsing in Rust" not sure if it's worth it

Thats true, and luckily that's not my goal with this feature request.
My goal is to start to replace some components of my data ingestion flow with Oxigraph equivalents.

I currently use RDFLib in Python (disclaimer; I am one of the maintainers of the RDFLib project). And because of the speed benefits of Hextuples in Python I often serialize intermediate RDF workflow artifacts from RDFLib using Hextuples. The first part of my workflow that I need to replace with Oxigraph needs to be able to ingest those Hextuples files into a Oxigraph in-memory dataset to use as a backend for running SHACL rules validation.

It would be relatively easy to switch to NQuads as my intermediate format to allow the use of Oxigraph in my workflow, but I figured no harm in raising the issue in this thread, to gauge interest in supporting Hextuples.

> If the goal is "there are hextuples files out there we need to load and they are not available in other formats"

I don't think that is the case. Hextuples is still a relatively young format, and you don't see published RDF files in the wild in that format. In my experience I see it used as an intermediate format, when exporting from one system to be ingested into another system. It works especially well when streamed through pipes.

**Comment by Tpt** [comment]
Thank you! Using NQuads seems indeed a good approach right now.

Out of curiosity, do you have an idea of the relative performance of hextuples vs nquads serialization in rdflib? This is only for my curiosity, if you don't have numbers don't waste time on this.

**Comment by ashleysommer** [comment]
No, I personally don't have any numbers to share. However one thing to note is rdflib recently got optional `orjson` support for using the simd-optimised C library to read JSON, and that is supported within the hextuples library.

I just did a quick test with a synthetic RDF dataset with 1 million quads, a mix of URIs and String literals. I wrote the same dataset out to both hextuples and nquads formats. The hextuples file is 205.7MB, the nquads file is 170.9 MB.

When parsing the file, the nquads parser in rdflib takes 15.6 seconds, the hextuples parser takes 12.6 seconds, and the hextuples parser with orjson enabled takes 11.1 seconds.
